### ISSUE TITLE
Followup to Solr integration -- installation steps

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#832]
Solr is now a requirement to run tests, which may increase barriers for newcomers. I'd like to take some steps to make this process easier:
- [x] update README to show solr installation: `apt-get install openjdk-7-jre openjdk-7-jdk icedtea-7-plugin`
- [x] remove duplicate rake db:setup and check if @icarito added those (or if we can remove them)
- [x] build `rake sunspot:solr:start` into `rake test` task until we separate it out
- [x] check cloud9 startup script to add java installation and solr startup lines: https://github.com/publiclab/plots2/blob/master/install_cloudnine.sh
- [x] see if we can turn off Solr dependencies for most tests, and separate out Solr tests to be optionally run under a `rake test:solr` task
- [ ] complete advanced search fixes in #800
- [ ] fix typeahead (#742)

**Comment by icarito** [comment]
I did not find the duplicate `rake db:setup` but it would be best to join both apt-get commands and drop the icedtea7-plugin which is not required

**Comment by jywarren** [comment]
ah, i already removed it and tested it, thanks -- maybe it resulted during
a rebase conflict. Oh, and somehow I didn't realize the icedtea7 plugin
wasn't required... must've mistakenly used that from a Stack Overflow
example. We should remove it from the Docker setup but also from the README
(i just updated that as well).

**Comment by jywarren** [comment]
In development mode, I'm getting:

```
Error:     {msg=SolrCore &apos;development&apos; is not available due to init failure: Index locked for write for core &apos;development&apos;. Solr now longer supports forceful unlocking via &apos;unlockOnStartup&apos;. Please verify locks manually!,trace=org.apache.solr.common.SolrException: SolrCore &apos;development&apos; is not available due to init failure: Index locked for write for core &apos;development&apos;. Solr now longer supports forceful unlocking via &apos;unlockOnStartup&apos;. Please verify locks manually!
```

I think this could be due to having started a second Solr instance for development, on top of the one I'm running for testing. This could be a common gotcha which we could build into the rake tasks...

**Comment by jywarren** [comment]
Yes -- that seems to have been the error -- and I had to shut **both** solr instances down and start up the right one before it worked.

**Comment by ananyo2012** [comment]
Ok seeing the discussion here the icedtea-7-plugin isn't required. But I am getting the same error as Jeff mentioned in the previous comment.

**Comment by ananyo2012** [comment]
Ah thanks Jeff! Sorry I didn't follow this before. I was having the same problem. Tests are running smoothly now. I think we should add this to the readme.

**Comment by ananyo2012** [comment]
@jywarren Should I remove the icedtea-7-plugin from the Dockerfile and Readme and update the readme with the test case as well?

**Comment by jywarren** [comment]
Yes please!

**Comment by ananyo2012** [comment]
Removed icedtea-7-plugin and updated the readme in #846

**Comment by jywarren** [comment]
Thanks!!!!

**Comment by ananyo2012** [comment]
@jywarren A lot of new contributors have come up in recently and they are facing problems in working up with the cloud9 installation. I can work on the cloud9 installation issues and building the rake tasks if you insist Jeff!

**Comment by ananyo2012** [comment]
@jywarren I was thinking if could somehow start and stop the solr server along with the rails server it would be great. Basically what we have to do is find a way to run a rake task while starting a stopping the rails server.

**Comment by ananyo2012** [comment]
@jywarren I am a bit stuck on this. Is there any way to find whether the solr server is already running? We have to put a condition to start the solr server in the test environment only if it is not already running. Otherwise rake is aborted due to the already running solr server. Similarly we need to stop the solr server in development only if it is running.

**Comment by jywarren** [comment]
Hmm, i'm not sure -- doing some searches for "detect solr running" -- but perhaps while you work on this, I could -- in parallel -- see if it's possible to skip Solr dependent tests and have a `rake test:nosolr` task for that.

**Comment by jywarren** [comment]
Unit tests may be omitted: http://stackoverflow.com/questions/2927405/how-to-skip-certain-tests-with-testunit#2940755

Looking for same with functional tests.

**Comment by jywarren** [comment]
[This](http://stackoverflow.com/questions/6744747/how-to-find-whether-solr-server-is-running-or-not?rq=1) is saying that you could just ping the solr service via wget, for example...

[This is saying](https://mail-archives.apache.org/mod_mbox/lucene-solr-user/201010.mbox/%3C4CC6ACA1.2020405@yahoo.de%3E) that you can check `/admin/ping` using wget. I mean, this'd assume that the client has wget to use rake, so that doesn't seem ideal...

**Comment by jywarren** [comment]
Alternatively, we could actually move the Solr tests in `searches_controller_test.rb` into `/test/solr/`, and run them with a separate `rake test:solr` command -- based on [this way of running a single test](http://stackoverflow.com/questions/274349/ddg#7986304):

`ruby -I test test/functional/whatevertest.rb`

**Comment by jywarren** [comment]
Notes:
* disable solr entirely in test mode: https://github.com/sunspot/sunspot/commit/b2dfd0e74d79c66915dafbdfff3c2ce1d364611b
* set Sunspot's session to a StubSessionProxy: http://stackoverflow.com/questions/5864470/disable-solr-during-database-rebuild (this could be used to make it not depend on the Solr service while running tests)
* ref'd in another issue, this is a gem allowing selective on/off of solr, though it's designed for RSpec: https://github.com/justinko/sunspot-rails-tester

**Comment by jywarren** [comment]
Whoa, i think setting `disabled:true` in sunspot.yml works -- i moved the `searches_controller_test` into its own folder, and will try to trigger that one with its own `rake test:sunspot`

**Comment by jywarren** [comment]
Got some more help here: http://flavio.castelli.name/2010/05/28/rails_execute_single_test/

Opened #861 which moves solr-dependent tests into a managed `rake test:solr` task, and adds it to the set run in `rake test:all` but not in `rake test`

**Comment by jywarren** [comment]
@icarito - is it possible to cache the java install, as that's failing in TravisCI pretty often. Although perhaps we should add priority to the plan to separate the Solr components into a separate docker instance, or use a pre-loaded one or something? I'm not familiar with how that works but could add it to this to-do list.
